File: problems/分治/53.最大子数组和.py
#!/usr/bin/env python3
"""
53. 最大子数组和（分治法）
"""
def max_sub_array_dc(nums: list[int]) -> int:
    def divide_conquer(nums, left, right):
        if left == right:
            return nums[left], nums[left], nums[left]
        mid = (left + right) // 2
        left_max, left_presum, left_sufsum = divide_conquer(nums, left, mid)
        right_max, right_presum, right_sufsum = divide_conquer(nums, mid+1, right)
        cross_max = left_sufsum + right_presum
        max_part = max(left_max, right_max)
        total_sum = left_max + right_max - min(left_max, right_max)
        return max(max_part, cross_max), max(left_presum, left_sufsum + right_presum), max(right_sufsum, left_sufsum + right_sufsum)
    return divide_conquer(nums, 0, len(nums)-1)[0]

if __name__ == "__main__":
    # 测试用例
    print(max_sub_array_dc([-2, 1, -3, 4, -1, 2, 1, -5, 4]))  # 预期: 6
    print(max_sub_array_dc([1]))  # 预期: 1
    print(max_sub_array_dc([5, 4, -1, 7, 8]))  # 预期: 23

File: problems/分治/53.最大子数组和.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Status {
    int sum, pref, suff, best;
};

Status combine(Status l, Status r) {
    Status res;
    res.sum = l.sum + r.sum;
    res.pref = max(l.pref, l.sum + r.pref);
    res.suff = max(r.suff, r.sum + l.suff);
    res.best = max({l.best, r.best, l.suff + r.pref});
    return res;
}

Status divideConquer(vector<int>& nums, int left, int right) {
    if (left == right) return {nums[left], nums[left], nums[left], nums[left]};
    int mid = (left + right) / 2;
    Status leftStatus = divideConquer(nums, left, mid);
    Status rightStatus = divideConquer(nums, mid+1, right);
    return combine(leftStatus, rightStatus);
}

int maxSubArrayDC(vector<int>& nums) {
    return divideConquer(nums, 0, nums.size()-1).best;
}

int main() {
    // 测试用例
    vector<int> nums1 = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    cout << maxSubArrayDC(nums1) << endl;  // 预期: 6
    vector<int> nums2 = {1};
    cout << maxSubArrayDC(nums2) << endl;  // 预期: 1
    vector<int> nums3 = {5, 4, -1, 7, 8};
    cout << maxSubArrayDC(nums3) << endl;  // 预期: 23
    return 0;
}
